src/primitives/poseidon.rs

```diff
@@ -0,0 +1,317 @@
+    /// to `[None; RATE]`.
+    type Rate: Default + AsRef<[Option<F>]> + AsMut<[Option<F>]>;
+
+    /// The arity of this specification.
+    fn arity() -> usize;
+
+    /// The number of full rounds for this specification.
+    fn full_rounds() -> usize;
+
+    /// The number of partial rounds for this specification.
+    fn partial_rounds() -> usize;
+
+    /// The S-box for this specification.
+    fn sbox(val: F) -> F;
+
+    /// Side-loaded index of the first correct and secure MDS that will be generated by
+    /// the reference implementation.
+    ///
+    /// This is used by the default implementation of [`Spec::constants`]. If you are
+    /// hard-coding the constants, you may leave this unimplemented.
+    fn secure_mds(&self) -> usize;
+
+    /// Generates `(round_constants, mds, mds^-1)` corresponding to this specification.
+    fn constants(&self) -> (Vec<Self::State>, Vec<Self::State>, Vec<Self::State>) {
+        let t = Self::arity();
+        let r_f = Self::full_rounds();
+        let r_p = Self::partial_rounds();
+
+        let mut grain = grain::Grain::new(SboxType::Pow, t as u16, r_f as u16, r_p as u16);
+
+        let round_constants = (0..(r_f + r_p))
+            .map(|_| {
+                let mut rc_row = Self::State::default();
+                for (rc, value) in rc_row
+                    .as_mut()
+                    .iter_mut()
+                    .zip((0..t).map(|_| grain.next_field_element()))
+                {
+                    *rc = value;
+                }
+                rc_row
+            })
+            .collect();
+
+        let (mds, mds_inv) = mds::generate_mds(&mut grain, t, self.secure_mds());
+
+        (
+            round_constants,
+            mds.into_iter()
+                .map(|row| {
+                    let mut mds_row = Self::State::default();
+                    for (entry, value) in mds_row.as_mut().iter_mut().zip(row.into_iter()) {
+                        *entry = value;
+                    }
+                    mds_row
+                })
+                .collect(),
+            mds_inv
+                .into_iter()
+                .map(|row| {
+                    let mut mds_row = Self::State::default();
+                    for (entry, value) in mds_row.as_mut().iter_mut().zip(row.into_iter()) {
+                        *entry = value;
+                    }
+                    mds_row
+                })
+                .collect(),
+        )
+    }
+}
+
+/// Poseidon-256 with arity 3, using the `x^5` S-box.
+#[derive(Debug)]
+pub struct P256Pow5T3<F: FieldExt> {
+    secure_mds: usize,
+    _field: PhantomData<F>,
+}
+
+impl<F: FieldExt> P256Pow5T3<F> {
+    pub fn new(secure_mds: usize) -> Self {
+        P256Pow5T3 {
+            secure_mds,
+            _field: PhantomData::default(),
+        }
+    }
+}
+
+impl<F: FieldExt> Spec<F> for P256Pow5T3<F> {
+    type State = [F; 3];
+    type Rate = [Option<F>; 2];
+
+    fn arity() -> usize {
+        3
+    }
```

Comment:
I need to verify whether these are correct for Poseidon-256; the paper is very unclear as to how many state and capacity words are in each of its specifications.

Reply:
The Orchard spec uses a rate of 2 and a capacity of 1, so we just need to figure out how that translates to their specifications in Table 1.

Reply:
Turns out I should have been looking at Table 2 in section 4.